pkg/config/context.go

```diff
@@ -34,5 +53,76 @@ func SyncContextsAndServers() error {
 			return errors.Wrap(err, "failed to set active context")
 		}
 	}
+
+	// After the sync, save the shas using the updated config
+	// so that next time, we avoid synching if the shas are the same
+	cfg, err = config.GetClientConfig()
+	if err == nil {
+		_ = saveContextAndServerShas(cfg)
+	}
 	return nil
 }
+
+func shouldSyncContextsAndServers(cfg *types.ClientConfig) bool {
+	var prevShas contextAndServerSha
+	err := datastore.GetDataStoreValue(lastContextServerShasKey, &prevShas)
+	if err != nil {
+		return true
+	}
+
+	newShas := computeShasForSync(cfg)
+	if newShas == nil || prevShas != *newShas {
+		return true
+	}
+	return false
+}
+
+//nolint:staticcheck // Deprecated
+func computeShasForSync(cfg *types.ClientConfig) *contextAndServerSha {
+	// tanzu contexts are not synced so we don't include them in the sha
+	// because even if they change, we don't need to sync them
+	var nonTanzuContexts []*types.Context
+	for i := range cfg.KnownContexts {
+		if cfg.KnownContexts[i].ContextType != types.ContextTypeTanzu {
+			nonTanzuContexts = append(nonTanzuContexts, cfg.KnownContexts[i])
+		}
+	}
+	ctxBytes, err := yaml.Marshal(nonTanzuContexts)
+	if err != nil {
+		return nil
+	}
+
+	// tanzu servers are not synced so we don't include them in the sha
+	// because even if they change, we don't need to sync them.
+	// Note that tanzu servers should normally not be in the known servers list
+	// but could happen for some older plugins or CLI versions which used to sync them.
+	var nonTanzuServers []*types.Server
+	for i := range cfg.KnownServers {
+		if cfg.KnownServers[i].Type != types.ServerType(types.ContextTypeTanzu) {
+			nonTanzuServers = append(nonTanzuServers, cfg.KnownServers[i])
+		}
+	}
+	serverBytes, err := yaml.Marshal(nonTanzuServers)
+	if err != nil {
+		return nil
+	}
+
+	return &contextAndServerSha{
+		ContextsSha: hashString(string(ctxBytes)),
+		// Only the k8s current context is synched
+		CurrentContextsSha: hashString(cfg.CurrentContext[types.ContextTypeK8s]),
```

Comment:
What about the `ContextTypeTMC`? Do we sync the current TMC context as well?

Reply:
In TPR, when we save a TMC context, we also save it on the config.yaml file as a server.
However, when the change the current TMC context, we don't sync this.
Therefore I only computed the SHA on the k8s context.

See https://github.com/vmware-tanzu/tanzu-plugin-runtime/blob/b5f34992ce053bc8979eec0a05260eaab828a193/config/contexts.go#L265-L266

Reply:
Ack. Thanks for confirming.